python/ray/train/mosaic/mosaic_trainer.py

```diff
@@ -0,0 +1,168 @@
+    ...         lr=0.05,
+    ...         momentum=0.9,
+    ...         weight_decay=2.0e-3,
+    ...     )
+    ...
+    ...     return composer.trainer.Trainer(
+    ...         model=model,
+    ...         train_dataloader=train_dataloader,
+    ...         optimizers=optimizer,
+    ...         **config
+    ...     )
+    ...
+    >>> scaling_config = ScalingConfig(num_workers=2, use_gpu=True)
+    >>> trainer_init_config = {
+    ...     "max_duration": "1ba",
+    ...     "train_dataset": train_dataset,
+    ...     "loggers": [InMemoryLogger()],
+    ...     "algorithms": [LabelSmoothing()],
+    ... }
+    ...
+    >>> trainer = MosaicTrainer(
+    ...     trainer_init_per_worker=trainer_init_per_worker,
+    ...     trainer_init_config=trainer_init_config,
+    ...     scaling_config=scaling_config,
+    ... )
+    ...
+    >>> trainer.fit()
+
+
+    Args:
+        trainer_init_per_worker: The function that returns an instantiated
+            ``composer.Trainer`` object and takes in an optional configuration
+            dictionary as an argument.
+        datasets: Any Ray Datasets to use for training.
+        trainer_init_config: Configurations to pass into ``trainer_init_per_worker`` as
+            kwargs.
+        torch_config: Configuration for setting up the PyTorch backend. If set to
+            None, use the default configuration. This replaces the ``backend_config``
+            arg of ``DataParallelTrainer``. Same as in ``TorchTrainer``.
+        scaling_config: Configuration for how to scale data parallel training.
+        dataset_config: Configuration for dataset ingest.
+        run_config: Configuration for the execution of the training run.
+        preprocessor: A ray.data.Preprocessor to preprocess the
+            provided datasets.
+        resume_from_checkpoint: A MosiacCheckpoint to resume training from.
+    """
+
+    def __init__(
+        self,
+        trainer_init_per_worker: Callable[[Optional[Dict]], Trainer],
+        *,
+        datasets: Optional[Dict[str, GenDataset]] = None,
+        trainer_init_config: Optional[Dict] = None,
+        torch_config: Optional[TorchConfig] = None,
+        scaling_config: Optional[ScalingConfig] = None,
+        dataset_config: Optional[Dict[str, DatasetConfig]] = None,
+        run_config: Optional[RunConfig] = None,
+        preprocessor: Optional["Preprocessor"] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+
+        self._validate_trainer_init_per_worker(
+            trainer_init_per_worker, "trainer_init_per_worker"
+        )
+
+        trainer_init_config = trainer_init_config.copy() if trainer_init_config else {}
+        if "_trainer_init_per_worker" in trainer_init_config:
+            raise ValueError(
+                "'_trainer_init_per_worker' is a reserved key in `trainer_init_config`."
+            )
+        trainer_init_config["_trainer_init_per_worker"] = trainer_init_per_worker
+
```

Comment:
can we just error if `datasets` is passed in for now.

Reply:
ValueError will be thrown when nonempty datasets is passed in